src/util/nonNull.ts

```diff
@@ -0,0 +1,43 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { isNullOrUndefined } from 'util';
+
+/**
+ * Retrieves a property by name from an object and checks that it's not null and not undefined.  It is strongly typed
+ * for the property and will give a compile error if the given name is not a property of the source.
+ */
+export function nonNullProp<TSource, TKey extends keyof TSource>(source: TSource, name: TKey): NonNullable<TSource[TKey]> {
+    const value: NonNullable<TSource[TKey]> = <NonNullable<TSource[TKey]>>source[name];
+    return nonNullValue(value, <string>name);
+}
+
+/**
+ * Validates that a given value is not null and not undefined.
+ */
+export function nonNullValue<T>(value: T | undefined, propertyNameOrMessage?: string): T {
```

Comment:
>nonNullValue [](start = 16, length = 12)

Nit: Since null and undefined are different, would it make more sense to make it verbose and specify nonNullOrUndedinedValue to be more explicit here? Same for other methods in this file.

Reply:
Nit: Also, looking at usages, we are essentially using side-effect of the getter to assert/throw... I personally would've probably broke the check + throw into a separate "ThrowIfNullOrUndefined" helper that would get used in both these getters and outside of the getters when you just want to check the value. Seems a bit cleaner to me, but not a big deal.

---
In reply to: [366155874](https://github.com/microsoft/vscode-azurearmtools/pull/443#discussion_r366155874) [](ancestors = 366155874)

Reply:
> > nonNullValue [](start = 16, length = 12)
> 
> Nit: Since null and undefined are different, would it make more sense to make it verbose and specify nonNullOrUndedinedValue to be more explicit here? Same for other methods in this file.

Yes, but I pulled it from another of our repros and was loathe to change the name. Also, since we don't normally handle null/undefined differently, the explicit/wordy naming seemed less important (this function is in other repos frequently used inside expressions, and the longer name makes the code harder to read).

Reply:
> Nit: Also, looking at usages, we are essentially using side-effect of the getter to assert/throw... I personally would've probably broke the check + throw into a separate "ThrowIfNullOrUndefined" helper that would get used in both these getters and outside of the getters when you just want to check the value. Seems a bit cleaner to me, but not a big deal.
> 
> In reply to: [366155874](https://github.com/microsoft/vscode-azurearmtools/pull/443#discussion_r366155874) [](ancestors = 366155874)

Related to the other question.  The main usage scenario in our other repos is this:
```
const myvalue = someexpression(nonNullProp(object, "propName"))
```

This was needed mostly because of some poorly-typed SDK objects.  I considered adding  assertIfNotX() but since they're completely equivalent decided not to bother.  This seems just as readable. But likewise no strong opinion.